chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,407 @@
+;
+; Announcements are used to ensure that each S_k follows the pattern is valid.
+; Announcements automatically commit to their own coin id.
+; Coin I_k creates an announcement that further commits to I_{k-1} and S_{k-1}.
+;
+; Coin I_k gets a proof that I_{k+1} is a cat, so it knows it must also create an announcement
+; when spent. It checks that I_{k+1} creates an announcement committing to I_k and S_k.
+;
+; So S_{k+1} is correct iff S_k is correct.
+;
+; Coins also receive proofs that their neighbors are cats, ensuring the announcements aren't forgeries, as
+; inner puzzles are not allowed to use `CREATE_COIN_ANNOUNCEMENT`.
+;
+; In summary, I_k generates an announcement Y_k (for "yell") as follows:
+;
+;  Y_k: hash of I_k (automatically), I_{k-1}, S_k
+;
+; Each coin ensures that the next coin's announcement is as expected:
+;  Y_{k+1} : hash of I_{k+1}, I_k, S_{k+1}
+;
+; TLDR:
+;  I_k : coins
+;  A_k : amount coin k contributes
+;  O_k : amount coin k spend
+;  D_k : difference/delta that coin k incurs (A - O)
+;  S_k : subtotal of debts D_1 + D_2 ... + D_k
+;  Y_k : announcements created by coin k commiting to I_{k-1}, I_k, S_k
+;
+; All conditions go through a "transformer" that looks for CREATE_COIN conditions
+; generated by the inner solution, and wraps the puzzle hash ensuring the output is a cat.
+;
+; Three output conditions are prepended to the list of conditions for each I_k:
+;  (ASSERT_MY_ID I_k) to ensure that the passed in value for I_k is correct
+;  (CREATE_COIN_ANNOUNCEMENT I_{k-1} S_k) to create this coin's announcement
+;  (ASSERT_COIN_ANNOUNCEMENT hashed_announcement(Y_{k+1})) to ensure the next coin really is next and
+;     the relative values of S_k and S_{k+1} are correct
+;
+; This is all we need to do to ensure cats exactly balance in the inputs and outputs.
+;
+; Proof:
+;   Consider n, k, I_k values, O_k values, S_k and A_k as above.
+;   For the (CREATE_COIN_ANNOUNCEMENT Y_{k+1}) (created by the next coin)
+;   and (ASSERT_COIN_ANNOUNCEMENT hashed(Y_{k+1})) to match,
+;   we see that I_k can ensure that is has the correct value for S_{k+1}.
+;
+;   By induction, we see that S_{m+1} = sum(i, 1, m) [O_i - A_i] = sum(i, 1, m) O_i - sum(i, 1, m) A_i
+;   So S_{n+1} = sum(i, 1, n) O_i - sum(i, 1, n) A_i. But S_{n+1} is actually S_1 = 0,
+;   so thus sum(i, 1, n) O_i = sum (i, 1, n) A_i, ie. output total equals input total.
+
+;; GLOSSARY:
+;;  MOD_HASH: this code's sha256 tree hash
+;;  GENESIS_COIN_CHECKER: the function that determines if a coin can mint new cats
+;;  INNER_PUZZLE: an independent puzzle protecting the coins. Solutions to this puzzle are expected to
+;;              generate `AGG_SIG` conditions and possibly `CREATE_COIN` conditions.
+;; ---- items above are curried into the puzzle hash ----
+;;  inner_puzzle_solution: the solution to the inner puzzle
+;;  prev_coin_id: the id for the previous coin
+;;  this_coin_bundle: ((parent_id puzzle_hash amount) . lineage_proof)
```

Comment:
Since the format of `this` and `next` are different, we should avoid confusion and not call them both `coin_bundle`. Come up with a new name for one of them.

Reply:
renamed to next_coin_info

Reply:
I'm not seeing this change.